### DISCUSSION TITLE
OpenSL ES is deprecated and replaced by Oboe

### DISCUSSION THREAD

**Discussion by juha-h** [baresip/baresip#3197] [Ideas]
Anyone else interested in writing baresip oboe module?  Migration guide from OpenSL ES is [here](https://github.com/google/oboe/blob/main/docs/OpenSLESMigration.md).

**Comment by sreimers** [comment]
I have [aaudio](https://developer.android.com/ndk/guides/audio/aaudio/aaudio) (oboe is only a C++ wrapper) on the roadmap

**Comment by juha-h** [comment]
Oboe falls back to OpenSL ES if AAudio is not available.  I don't know if most Android devices now support AAaudio.  If not, oboe module would always work.

**Comment by juha-h** [comment]
AAaudio page tells:
>Note: Developers should consider using the open source Oboe library which is available on [GitHub](https://github.com/google/oboe). Oboe is a C++ wrapper that provides an API that closely resembles AAudio. It calls AAudio when it is available, and falls back to [OpenSL ES](https://developer.android.com/ndk/guides/audio/opensl) if AAudio is not available.

**Comment by juha-h** [comment]
Here is sample code that looks useful when writing the baresip module: https://github.com/google/oboe/tree/main/samples/LiveEffect/src/main/cpp

**Comment by sreimers** [comment]
We always try to use C APIs if available, it's much easier for implementation. The only benefit I see is the OpenSL ES fallback, but `AAudio` is available since Android Oreo (2017).

**Reply by juha-h** [reply]
OK, AAudio then.

**Comment by juha-h** [comment]
Any C usage example somewhere?  All that I have so far found use C++.

**Comment by juha-h** [comment]
The current minimum API level of my baresip app is 23 (Andoid 6).  Oreo is 26 (Android 8).  Using AAudio instead of Oreo would mean that Android 6 and 7 would not anymore be supported.

**Comment by sreimers** [comment]
This looks fine to me, since this is a new module. Using a modern app on >8 Year Android Release is nothing I would worry about, since users can also use a old baresip android version instead.

![image](https://github.com/user-attachments/assets/9d014594-4366-4810-9605-8c470d623678)

https://composables.com/android-distribution-chart

**Comment by juha-h** [comment]
Sebastian Reimers writes:

> This looks fine to me, since this is a new module. Using a modern app
> on >8 Year Android Release is nothing I would worry about, since users
> can also use a old baresip android version instead.

OK.

**Comment by juha-h** [comment]
While reading the docs, I noticed that that when stream is built it is possible to use `setSessionId` function to  `Allocate SessionID to connect to the Java AudioEffects API`.  AudioEffects API includes stuff like `AcousticEchoCanceler` and `AutomaticGainControl`.  Perhaps then there would be no need anymore for `webrtc_aecm` and `augain` modules.

**Comment by juha-h** [comment]
Since AAudio goes way above my know-how, I started to play with implementing oboe module: https://box.tutpro.com/tmp/oboe/ just to check if it is possible to get access to `sesssionId`s of the streams.  And indeed it was:
```
11-13 18:33:32.451  8395  8452 D Baresip Lib: oboe: session id, channel count, sample rate = 8105, 1, 16000
```
So it might then be possible to pass those to the app using `UA_EVENT_MODULE` and attached aec and gain audio effects to the streams.

I haven't yet properly implemented `onAudioReady` callback that handles the audio data.

**Comment by juha-h** [comment]
I added to oboe module sending of module events that tell the app the `sessionid`s and then handled recorder sessionid event like this:
```
        if (ev[0] == "recorder sessionid") {
            val sessionId = ev[1].toInt()
            Log.d(TAG, "got recorder sessionid $sessionId")
            if (AcousticEchoCanceler.isAvailable()) {
                Log.i(TAG, "AcousticEchoCanceler IS available")
                AcousticEchoCanceler.create(sessionId)
            } else {
                Log.i(TAG, "AcousticEchoCanceler IS NOT available")
            }
            if (AutomaticGainControl.isAvailable()) {
                Log.i(TAG, "AutomaticGainControl IS available")
                AutomaticGainControl.create(sessionId)
            } else {
                Log.i(TAG, "AutomaticGainControl IS NOT available")
            }
            return
        }
```
That produced to log:
```
11-14 10:36:54.447  8809  8867 D Baresip Lib: oboe: session id, channel count, sample rate = 8865, 1, 16000
11-14 10:36:54.459  8809  8866 I Baresip : AcousticEchoCanceler IS available
11-14 10:36:54.475  8809  8866 I Baresip : AutomaticGainControl IS available
11-14 10:36:54.507  8809  8867 D Baresip Lib: audio: source started with sample format S16LE
```
Unbelievable if implementing aec and gain control would be so easy.  From the docs I got the impression that aec and gain control are only applied applied to recorder sessionid (not to player sessionid).

Proper handling of `onAudioReady` callbacks is still missing from oboe module and I'm not sure if I know how to do it.

**Comment by juha-h** [comment]
Oboe API has this:
```
class AudioStreamDataCallback {
public:
    virtual ~AudioStreamDataCallback() = default;
virtual DataCallbackResult onAudioReady(
            AudioStream *audioStream,
            void *audioData,
            int32_t numFrames) = 0;
};
```
Doc tells: `For an output stream, onAudioReady function should render and write numFrames of data  in the stream's current data format to the audioData buffer.`

Haven't figured out, how it can get the data from baresip modules player.

**Comment by juha-h** [comment]
I now created start of aaudio module: https://box.tutpro.com/tmp/aaudio.  Otherwise complete, but all actual audio processing (callbacks) is still missing :-). Docs:

https://developer.android.com/ndk/guides/audio/aaudio/aaudio
https://developer.android.com/ndk/reference/group/audio#group___audio_1ga515e4769d8524132ca1e560020d0434d

Should I create a branch to baresip project in case other people would like to help?

**Comment by juha-h** [comment]
I just pushed `aaudio` branch to baresip repo.  The module compiles, but handling of the audio data may not be correct.  Especially I would like to get review of `dataCallback` function in `player.c` and `receiver.c`.  The comments in front of both tell what the callback function should do.  I have copied most of the callback code from pulse module.

**Comment by juha-h** [comment]
I tried the module on my Android device and there was understandably audio both directions, but very, very glitchy.  So obviously something is still wrong in audio processing.  Any help would be appreciated.
